examples/pytorch/dino.py

```diff
@@ -0,0 +1,90 @@
+import torch
+from torch import nn
+import torchvision
+import copy
+
+from lightly.data import LightlyDataset
+from lightly.data import DINOCollateFunction
+from lightly.loss import DINOLoss
+from lightly.models.modules import DINOProjectionHead
+from lightly.models.utils import deactivate_requires_grad
+from lightly.models.utils import update_momentum
+
+class DINO(torch.nn.Module):
+    def __init__(self, backbone, input_dim):
+        super().__init__()
+        self.student_backbone = backbone
+        self.student_head = DINOProjectionHead(input_dim, 512, 64, 2048)
+        self.teacher_backbone = copy.deepcopy(backbone)
+        self.teacher_head = DINOProjectionHead(input_dim, 512, 64, 2048)
+        deactivate_requires_grad(self.teacher_backbone)
+        deactivate_requires_grad(self.teacher_head)
+    
+    def forward(self, x):
+        y = self.student_backbone(x).flatten(start_dim=1)
+        z = self.student_head(y)
+        return z
+
+    def forward_teacher(self, x):
+        y = self.teacher_backbone(x).flatten(start_dim=1)
+        z = self.teacher_head(y)
+        return z
+
+resnet = torchvision.models.resnet18()
+backbone = nn.Sequential(*list(resnet.children())[:-1])
+input_dim = 512
+# instead of a resnet you can also use a vision transformer backbone as in the
+# original paper (you might have to reduce the batch size in this case):
+# backbone = torch.hub.load('facebookresearch/dino:main', 'dino_vits16', pretrained=False)
```

Comment:
Would add an example using timm models (see https://github.com/rwightman/pytorch-image-models) here in the comments. Either additionally or instead of the facebook research one. Just simply because the repo has tons of more models and is well maintained :)

Reply:
I just tested it and the vision transformers in timm expect all input images to have a fixed size. But DINO makes global and local crops with various sizes :(